### PR TITLE
service: Don't enter failed state if we will auto restart

### PR DESCRIPTION
**Pull request by daandemeyer**
We document in systemd.unit that services using `Restart=` enter the failed state only after the start limits are reached. Let's make sure we actually do that by not switching to the failed state before going to the auto restart state.

### CODE REVIEW

**Comment by poettering** [comment]
This was actually a concious design decision back in the day, see the preceeding comments in the sources for example, or the git discussion of various commits earlier that touch this.

People look for PropertiesChanged events via the bus for example expecting that if there's a failure condition they see the failure state popping up.

I must say I find the existing logic a bit icky from today's pov, but I doubt we can change it too drastically, simply because it is API.

Maybe a compromise might be to transition through a substate here that indicates what is going on.

i.e. right now the unit state UNIT_FAILED maps to the service substate SERVICE_FAILED, and UNIT_DEAD to SERVICE_DEAD. very simple. I'd change this so that in addition to SERVICE_FAILED and SERVICE_DEAD we'd have SERVICE_FAILED_BEFORE_AUTO_RESTART and SERVICE_DEAD_BEFORE_RESTART. Both would also map to the unit state UNID_FAILED/UNIT_DEAD however. 

So, in other words: on the lower level we'd distinguish the failed case which we'll stay in from the failed case we'll just quickly transition through. On the high-level this would not be distinguished however.

Strictly speaking such a low-level state rearrangement could also be considered a compat break, but so far we've been a bit more liberal regarding changes to the low-level state engine compared to the high-level one.

**Comment by DaanDeMeyer** [comment]
Closing in favor of #26751

**Comment by poettering** [comment]
I implemented the state split idea in #26971 btw.
